docs/architecture/adr-031-msg-service.md

```diff
@@ -0,0 +1,216 @@
+actual method routing, allowing this feature to be added incrementally on top of
+existing functionality.
+
+### `ServiceMsg` interface
+
+All request messages will need to implement the `MsgRequest` interface which is a
+simplified version of `Msg`, without `Route()`, `Type()` and `GetSignBytes()` which
+are no longer needed:
+
+```go
+type MsgRequest interface {
+  proto.Message
+  ValidateBasic() error
+  GetSigners() []AccAddress
+}
+```
+
+`ServiceMsg` will forward its `ValidateBasic` and `GetSigners` methods to the `MsgRequest`
+methods.
+
+### Module Configuration
+
+In [ADR 021](./adr-021-protobuf-query-encoding.md), we introduced a method `RegisterQueryService`
+to `AppModule` which allows for modules to register gRPC queriers.
+
+To register `Msg` services, we attempt an more extensible approach by converting `RegisterQueryService`
+to a more generic `RegisterServices` method:
+
+```go
+type AppModule interface {
+  RegisterServices(Configurator)
+  ...
+}
+
+type Configurator interface {
+  QueryServer() grpc.Server
+  MsgServer() grpc.Server
+}
+
+// example module:
+func (am AppModule) RegisterServices(cfg Configurator) {
+	types.RegisterQueryServer(cfg.QueryServer(), keeper)
+	types.RegisterMsgServer(cfg.MsgServer(), keeper)
+}
+```
+
+The `RegisterServices` method and the `Configurator` interface are intended to
+evolve to satisfy the use cases discussed in [\#7093](https://github.com/cosmos/cosmos-sdk/issues/7093)
+and [\#7122](https://github.com/cosmos/cosmos-sdk/issues/7421).
+
+When `Msg` services are registered, the framework _should_ verify that all `Msg...Request` types
+implement the `MsgRequest` interface described above and throw an error during initialization rather
+than later when transactions are processed.
+
+### `Msg` Service Implementation
+
+Just like query services, `Msg` service methods can retrieve the `sdk.Context`
+from the `context.Context` parameter method using the `sdk.UnwrapSDKContext`
+method:
+
+```go
+func (k Keeper) SubmitProposal(goCtx context.Context, params *types.MsgSubmitProposal) (*MsgSubmitProposalResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+    ...
+}
+```
+
+The `sdk.Context` should have an `EventManager` already attached by the `ServiceMsg`
+router.
+
+Separate handler definition is no longer needed with this approach.
+
+## Consequences
+
+### Pros
+- communicates return type clearly
+- manual handler registration and return type marshaling is no longer needed, just implement the interface and register it
+- some keeper code could be automatically generate, this would improve the UX of [\#7093](https://github.com/cosmos/cosmos-sdk/issues/7093) approach (1) if we chose to adopt that
+- generated client code could be useful for clients
+
+### Cons
+- supporting both this and the current concrete `Msg` type approach simultaneously could be confusing
+(we could choose to deprecate the current approach)
```

Comment:
I would +1 this, we could even rename it to `sdk.LegacyMsg`.

I am still not clear where to put the `GetSignBytes()` method in this case though.

Reply:
`GetSignBytes` is only needed for amino. So `LegacyMsg` would make sense for that.